### PR TITLE
Add XMLHttpRequestUpload support

### PR DESCRIPTION
**Pull request by snstanton**
* factor out MockEventTarget base class
* add upload property to MockXMLHttpRequest
* add uploadProgress() method to MockRequest
* update readme to reflect new api
* fix readme to describe .progress() on the proper object

### CODE REVIEW

File: lib/MockEventTarget.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+}
+
+
+/**
+ * Trigger an event
+ * @param   {String} event
+ * @param   {Object} eventDetails
+ * @returns {MockEventTarget}
+ */
+MockEventTarget.prototype.trigger = function(event, eventDetails) {
```
Comment (jameslnewell) [inline lib/MockEventTarget.js]: should this be `dispatchEvent`?

File: lib/MockXMLHttpRequest.js [hunk 2]
```diff
@@ -171,6 +165,8 @@ MockXMLHttpRequest.prototype.send = function(data) {
         }, typeof(timeout) === 'number' ? timeout : self.timeout+1);
 
       } else {
+        //trigger a load event to indicate the data has been sent
+        self.upload.trigger('load');
```
Comment (jameslnewell) [inline lib/MockXMLHttpRequest.js]: `onloadstart` too? https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/upload

we should probably dispatch the other events e.g. `onloadend` at some stage too?

File: README.md [hunk 3]
```diff
@@ -134,6 +134,14 @@ Get the request headers.
 
 Get the request body.
 
+#### .progress(loaded : number, total : number, lengthComputable : bool)
```
Comment (jameslnewell) [inline README.md]: thanks for the fix here. do you think this should have been implemented on the response interface?

Comment (jameslnewell) [inline README.md]: what do you think about 
```
req.progress(...) //dispatches xhr.upload.onprogress
res.progress(...) //dispatches xhr.onprogress
```

**Comment by jameslnewell** [comment]
I've merged this into [v2](https://github.com/jameslnewell/xhr-mock/tree/v2).

File: lib/MockEventTarget.js (hunk 1 above)
Comment (snstanton) [inline lib/MockEventTarget.js]: trigger seems clear enough

File: lib/MockXMLHttpRequest.js (hunk 2 above)
Comment (snstanton) [inline lib/MockXMLHttpRequest.js]: possibly so.  I didn't need any of the other events to get my tests working, but having a full lifecycle would probably be useful.

File: README.md (hunk 3 above)
Comment (snstanton) [inline README.md]: That's a good suggestion.  I like the symmetry of having the upload related events on the request and the download related events on the response.

**Comment by joscha** [comment]
@jameslnewell any chance we could get these changes in a release, so we can use `xhr.upload.addEventListener` ?

**Comment by jameslnewell** [comment]
@joscha I've fixed the last few red tests and released a preview release of v2 (`xhr-mock@next` or `xhr-mock@2.0.0-preview.1`). There's a few TODO/FIXME items left in order to get complete compliance with the spec, but v2 should already be much more compliant than v1. If you could start using/testing v2 for me, I'll release it after a couple weeks of no bugs being reported.

Thanks for your patience.

**Comment by joscha** [comment]
Thanks @jameslnewell

**Comment by itsjamie** [comment]
@jameslnewell can you publish with the lib folder for the preview build? Currently the "lib" folder doesn't exist in the npm package that is published at `xhr-mock@next`.

**Comment by jameslnewell** [comment]
@itsjamie Sorry! Try `xhr-mock@2.0.0-preview.2`.

/cc @joscha

**Pull request closed without merging**